---
title: "Pituitary corticotroph gene expression analysis"
subtitle: "Cheung et al scRNA seq dataset"
author: "Craig McDougall"
date: "October 2020"
output: html_document
#output: 
# html_document:
#    toc: true
#output: pdf_document
#output: ioslides_presentation
  
---

```{r setup,  include=FALSE}
#echo r code in markdown output,
knitr::opts_chunk$set(echo = TRUE)
 
```

## Introduction

This document details the analysis of the single RNA sequence data by [Chung et.al. 2018](https://academic.oup.com/endo/article/159/12/3910/5133692). available from [the NCBI gene expression omnibus](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE120410).


## Step 1. Call the relevant Libraries

Before the main boy of the pipeline is run, all of the required libraries are called.

```{r Initialising required libraries, message=FALSE}

library(dplyr) 
library(Seurat)
library(patchwork)
library(sctransform)
library(ggplot2)
library(scales)
```

## Step 2. Loading data and initialising Seurat object.

The raw count and gene data is read in from the 10X sequence data files downloaded from the public repositories. Seurat expects "barcodes.tsv", "matrix.tsv" and "matrix.tx" files. These are then used to initialise a Seurat object for the transcriptome analysis.


```{r Loading in the Data & Initialising Seurat, echo=FALSE}
#load raw data using Read10x 
raw.data <- Read10X("Cheung Data")

#initialise a Seurat object with the raw (not normalised) data. This is a count matrix
data <- CreateSeuratObject(counts=raw.data, project = "data", min.cells = 3, min.features = 200) 
```

## Step 3. Quality control plots

Before processing the data, some quality control plots were generated to assess the quality of the data;
```{r Data visualisation/Quality control, echo = FALSE}
#Seurat recommends percent.MT label for mitochondria data, however;
#Cheung, Ho & Mayran et.al. use "percent.mt"
data <- PercentageFeatureSet(data, pattern = "^mt-", col.name = "percent.mt")
#Fletcher uses "percent.Mt"
#data <- PercentageFeatureSet(data, pattern = "^mt-", col.name = "percent.MT")
#select appropriate method and comment out other option

#use QC metrics to filter the cells
#visualise the QC metrics - use a violin plot 
VlnPlot(data, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

#make the plots
plot1 <- FeatureScatter(data, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(data, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
#then plot
plot1 + plot2

```

## Step 4.  Transform the data

[SCTransform](https://satijalab.org/seurat/v3.1/sctransform_vignette.html) is used to filter, scale and regress out any confounding factors and normalise the data. It also identifies variable genes.
 

``` {r Perform Filtering, Normalisation, Regression and variable gene assessment, echo=FALSE }
data <- SCTransform(data,vars.to.regress = "percent.mt")
```

## Step 5. Dimensional reduction

PCA is used to transform the data from the table into new features known as principal components, which capture the information of the dataset in a new way.

```{r Perform PCA, echo=FALSE}
data <- RunPCA(data,verbose= FALSE) 

```

An elbow plot is used to assess which principal components contain the majority of the information and identify which may be removed from the dataset;

``` {r Plot Elbow, echo = FALSE }
#Do we need to keep all PCs, or is majority of data captured by a certain point?
ElbowPlot(data, ndims=30)

```


Uniform Manifold Approximation and Projection (UMAP) dimensional reduction is performed on the principal components.

``` {r Perform Dimensional reduction (UMAP), eval=FALSE }
#UMAP dimensional reduction of the identified PCs
data <- RunUMAP(data, dims=1:30, umap.method = "umap-learn", metric = "correlation", verbose=FALSE) 

```



## Step 6. Clustering

First, shared nearest neighbour (SNN) information is calculated, before identifying clusters. The clusters are then projected into two lower dimensional space using a [dimensional reduction plot](https://www.rdocumentation.org/packages/Seurat/versions/3.1.4/topics/DimPlot) for visualisation.

```{r Perform clustering, echo=FALSE}
#Find nearest neighbours
data <- FindNeighbors(data, dims=1:30, verbose=FALSE)

#CLuster based on neighbour distances
data <- FindClusters(data, verbose=FALSE)

#plot the clusters
DimPlot(data, label=TRUE) + NoLegend()

```
Once the clusters have been established, they can be projected into two lower dimensional space using a [dimensional reduction plot](https://www.rdocumentation.org/packages/Seurat/versions/3.1.4/topics/DimPlot) for visualisation.



## Step 7. Cell Type assessment
The main question now, is how well do the clusters approximate cell type? It is difficult to make absolute claims when attributing cell types to empirical cluster data based purely on an *in silico* analysis. However we may approximate the cell type of each cluster based on known canonical expression markers ^[[Fletcher et.al 2019](https://www.frontiersin.org/articles/10.3389/fendo.2019.00623/full)]^ ^[[Chung et.al 2018](https://academic.oup.com/endo/article/159/12/3910/5133692)]^. 


In the corticotroph study, hormone secreting pituitary cell types were identified based on the following canonical gene expression markers: 

 * Thyrotrophs (Pou1f1 + Tshb),
 * Somatotrophs(Pou1f1 + Gh)
 * Lactotrophs (Pou1f1 + Prl)
 * Gonadotrophs (Lhb)
 * Melanotrophs (Pomc + Pcsk2 + Pax7)
 * Corticotrophs (Pomc + Crhr1 + Avpr1b + Gpc5 - Pcsk2 - Pax7)
 
 [Feature plots](https://www.rdocumentation.org/packages/Seurat/versions/3.1.4/topics/FeaturePlot) were used to visually assess gene expression levels of single cells. 

``` {r Perform cell type assessment, eval=FALSE}
#Visualise the canonical marker genes 

#first exclude Pou1f1 expressing cells (Thyrotrophs, Somatotrophs, Lactotrophs)
FeaturePlot(data, features = c("Pou1f1","Tshb", "Gh","Prl"), pt.size = 0.2)&scale_color_viridis_c() 

#then exclude the gonadotrophs
FeaturePlot(data, features = c("Lhb"), pt.size = 0.2)&scale_color_viridis_c() 

#then exclude the melanotrophs
FeaturePlot(data, features = c("Pomc","Pax7","Pcsk2"), pt.size = 0.2)&scale_color_viridis_c() 

#then id the corticotrophs
FeaturePlot(data, features = c("Pomc", "Crhr1","Avpr1b","Gpc5"), pt.size = 0.2)&scale_color_viridis_c() 
```
The first plot would be used to exclude clusters of thyrotrophs, Somatotrophs and lactotrophs. The second plot allows exclusion on gonadotrophs, the third melanotrophs. The final plot allows confirmation of the corticotoph cluster.

![Attributing cell type may be estimated by canonical gene expression markers, here the cluster of corticotroph cells may be estimated as cluster 14](/Users/craigmcdougall/Documents/Project 2/Data Analysis Reporting/feature plot.png){width=50%}


## Step 8. Isolate desired cell cluster

To examine cell homogeneity within the corticotrophs, the cluster identified in the previous step was pulled out as a subset. 

The desired cluster must be manually entered upon visual inspection of the gene expression feature plots in the previous step.
``` {r isolate desired cells, eval=FALSE}
#enter the cotricoph cell ids, found in visualisation. Note change to correct cluster #s or will miss-match
#new.cluster.ids <- c("0", "1", "2", "3", "4", "5", "6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22")
new.cluster.ids <- c("0", "1", "2", "3", "4", "5", "6","7","8","9","10","11","12","13","cort","15","16","17","18","19","20","21","22")

names(new.cluster.ids) <- levels(data)

data <- RenameIdents(data, new.cluster.ids)
#then put on the plot to confirm slected the correct one (sanity check)
DimPlot(data, reduction ="umap", label = TRUE, pt.size = 0.5)+NoLegend()
#or pull out a  seurat of cort only
cortico <- subset(data, idents = "cort")
```
As a sanity check that the correct cluster was selected, a dimensional reuction plot

![The corticotroph data is isolated and a sanity check performed to ensure the correct cluster has indeed been selected](/Users/craigmcdougall/Documents/Project 2/Data Analysis Reporting/cort.png){width=50%}

## Step 9. Re-cluster selected cell cluster

To investigate cell homogeneity within the cell cluster of interest, the same procedure may be repeated;

*Data transformation, identification of principal components (PCA)
*Dimensional reduction (UMAP)
*Identification of nearest neighbours (SNN)
*Clustering using SNN.

``` {r re-cluster selected cluster, eval=FALSE}

#dimensional analysis
#PCA to identify PCs
cortico <- RunPCA(cortico,verbose= FALSE) 

#dimensional reduction
#UMAP dimensional reduction of the identified PCs
cortico <- RunUMAP(cortico, dims=1:30, verbose=FALSE) #note because SCTransform passes 3000 features can run more PCs

#Find nearest neighbours
cortico <- FindNeighbors(cortico, dims=1:30, verbose=FALSE)

#CLuster based on neighbour distances
cortico <- FindClusters(cortico, verbose=FALSE)

#plot the clusters
DimPlot(cortico, pt.size = 5, label=TRUE) + NoLegend()

#Visualise the canonical marker genes 
FeaturePlot(cortico, features = c("Pomc"), pt.size = 5, label=TRUE)& scale_color_viridis_c() 

#examine DE genes in the  corticotroph cluster
#find the markers for EVERY CLUSTER when compared to remaining cells, only +ve ones
cortico.markers <-FindAllMarkers(cortico, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
#examine top 2 deferentially expressed genes in each corticotroph cluster
cortico.markers %>% group_by(cluster) %>% top_n(n=2, wt= avg_logFC)

```

A dimensional reduction plot can then reveal any heterogeneity in the cluster;

![A re-clustering of the corticotroph data may reveal cell heterogeneity](/Users/craigmcdougall/Documents/Project 2/Data Analysis Reporting/subcluster.png){width=50%}

A featureplot can be used to examine gene expression levels across these clusters within the corticotroph data;

![A featureplot can illustrate gene expression levels across these clusters within the corticotroph data, for example Pomc expression is higher in cluster 0 & 2.](/Users/craigmcdougall/Documents/Project 2/Data Analysis Reporting/pomc.png){width=50%}




Differential expression analysis can reveal differences between the clusters, for example the two top deferentially expressed genes for each cluster are;
```{r DE table, echo=FALSE}

DE <- read.csv("/Users/craigmcdougall/Documents/Project 2/Data Analysis Reporting/DE.csv")
head(DE)
```

## Summary

The described pipeline is a powerful tool to investigate gene expression and cell homogeneity within sub-populations, however it is just a tool. For meaningful results to be obtained the user must apply knowledge of the samples under study to ensure biologically sensible data is being studied and any decisions based on sound biological features. The output is only as good as the input and parameters.

## Future improvements
A future version of this script will include cell cycle considerations. 

All suggestions for future improvements are welcome, as are comments on the current version. Please make these via the [GitHub](https://github.com/cmcdougall5/scRNAseq_pipeline) page.





    