---
title: "Single Cell RNA Sequence Analysis Pipeline Guide"
author: "Craig McDougall"
date: "October 2020"
output: html_document
#output: 
 #html_document #:
    #toc: true
#output: pdf_document
#output: ioslides_presentation
  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
 
```

## Introduction

This guide describes a bioinformatic pipeline, designed in R to analyse single cell RNA sequence data. The pipeline was constructed as part of an MSc project at the University of Edinburgh in Scotland. The project was investigating sex differences in the central stress response system; the Hypothalmic-pituitary-adrenal (HPA) axis. The pipeline was used to probe the transcriptome profiles of anterior pituitary cells, concentrating on corticotrophs, which were the focus of the sex differences study.  

The pipeline files can be found on [GitHub](https://github.com/cmcdougall5/scRNAseq_pipeline).

The pipeline is based on the [Seurat R. package](https://satijalab.org/seurat/), in particular the [guided clustering tutorial](https://satijalab.org/seurat/v3.2/pbmc3k_tutorial.html), but updated to include the `SCTransform` normalisation function.  

Example single cell RNA sequence data is available in public repositories. In the project the pipeline was designed for, the following data sets were examined:

*   [Chung et al](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE120410). 
*   [Mayran et al](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi). 
*   [Ho et al](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi). 
*   [Fletcher et al](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE132224). 

Details on the data acquisition is described within the following publications:

*  [Chung et al](https://academic.oup.com/endo/article/159/12/3910/5133692). 
*  [Mayran et al](https://www.nature.com/articles/s41467-019-11791-9). 
*  [Ho et al](https://link.springer.com/article/10.1007/s13238-020-00705-x). 
*  [Fletcher et al](https://www.frontiersin.org/articles/10.3389/fendo.2019.00623/full). 

The pipeline may be broken down into the following steps.

## Precursor: Setting directories

As a precursor to the main body of the pipeline script, the data directories for the location of input data and output data storage locations are set and verified;


`
#set the working directory to directory wish to save analysis results to
setwd("/Users/craigmcdougall/Documents/Project 2/Data analysis/Cheung")

#check working directory
getwd()


#define the data directory where the data is held
data.dir <- "/Users/craigmcdougall/Documents/Project 2/Data/Cheung"

#list all the files in the data directory: barcodes, genes, matrix
list.files(data.dir)
`

The correct path must be specified in order for the script to run correctly, hence the sanity checks to ensure the curreent working directory is the intended one.

With the initial setup verified the script moves onto the next step in the pipeline.

## 1. Call the relevant Libraries

Before the main boy of the pipeline is run, all of the required libraries are called;

```{r Initialising required libraries, eval = FALSE}

library(dplyr) 
library(Seurat)
library(patchwork)
library(sctransform)
library(ggplot2)
library(scales)
```

## 2. Loading data and initialising Seurat object.

The raw count and gene data is read in from the 10X sequence data files downloaded from the public repositories. Seurat expects "barcodes.tsv", "matrix.tsv" and "matrix.tx" files. These asre then used to initialise a Seurat object for the transcriptome analysis.


```{r Loading in the Data & Initialising Seurat, eval=FALSE}
#load raw data using Read10x 
raw.data <- Read10X(data.dir=data.dir)

#initialise a Seurat object with the raw (not normalised) data. This is a count matrix
data <- CreateSeuratObject(counts=raw.data, project = "data", min.cells = 3, min.features = 200) 
```

## 3. Quality control plots

Before processing the data, some quality control plots are made to assess the quality of the data and explore confounding factors such as assessing the number of reads that map to mitochondrial genome.

```{r Data visualisation/Quality control}
#Seurat recommends MT- prefix for mitochondria data, however
#cheung, ho, mayran use mt
data <- PercentageFeatureSet(data, pattern = "^mt-", col.name = "percent.mt")
#Fletcher uses Mt-
#data <- PercentageFeatureSet(data, pattern = "^mt-", col.name = "percent.MT")

#use QC metrics to filter the cells
#visualise the QC metrics - use a violin plot 
VlnPlot(data, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

#make the plots
plot1 <- FeatureScatter(data, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(data, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
#then plot
plot1 + plot2
```


















This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

## Text Formatting Basics

### Headers

# First Level Header

## Second Level Header

### Third Level Header

If you don't want to render as header, wrap with backticks
'#First level header'

### Bulleted and Numbered Lists

* List element 1
* List element 2
* List item 3a
* List item 3b

1. Numbered list 1
1. Numbered list 2
    * Item 1
    * Item 2
    
### Text Formatting 

* Make text italic like *this* or _this_
* Make text bold like **this** or __this__
* Use `backticks` for code.

* Wrap a character to subscript in tildes (`~`). For example, `H~2~O` renders as H~2~O.
* Wrap a character to superscript in carets (`^`), like this `R^2^` renders R^2^.


### Links

Direct in-line links: <https://rmarkdown.rstudio.com/>

Phrase links: RStudio's [R Markdown page](https://rmarkdown.rstudio.com/).



### Code Chunks

To run blocks of code in R Markdown, use code chunks. Insert a new code chunk with:
1. `Command + Option + I` on a Mac
2. Another option is the "insert" drop-down icon in the toolbar and selecting R.

### Running Code

R studio provides many options for running code chinks in the "Run" drop-down tab on the toolbar.

Before running code chunks it is often a good idea to restart R session ans start with a clean environment. Do this with `Command+Shift+F10`.

Shortcuts to run code
1. Run all chinks above the current chunk with `command+option+p`.
1. Run the current chunk with `command+Option+N`
1. Run the next chunk with `command+option+R`
1. Run all chunks with `command+option+R` or `command+A+Enter`

### Control Behaviour with Code Chunk Options

Many options how to control each piece of code.

* `echo = FALSE`: Do not show code in the output, but run the code and produce all the outputs, lots, warnings and messages. 
* `eval = FALSE`: Show the code, but do not evaluate the code
* `fig.show="hide"`:Hide the plots
* `include = FALSE`: Run code, but suppress all output. 
* `message = FALSE`: Prevent packages from printing messages when they load and messages generated by functions.

### Inline code
Directly embed R code into an R Markdown document with inline code. Useful when including information about your data in written summary.

Use `r` with backticks:  The `cars` dataset contains r nrow(cars) rows and r ncol(cars) columns. Would look like:


Use `r` with backticks:  The `cars` dataset contains `r nrow(cars)` rows and `r ncol(cars)` columns.

Can reduce errors by summarising information programmatically.

### Navigating sections and code chunks

It is useful to name code chunks in long documents, name like:
```{r my_boring_chunk_name}
nrow(cars)
```

Can then use the navigator at the bottom of the pane or use in other sections of document.

### Table formatting

By default tables displayed as you see them in the R console. Can improve by using knitr::kable(). eg:
```{r table formatting}
knitr::kable(head(cars), 
caption = "The first Few rows of cars dataset")
```

### Output format options

Meta data in the YAML controls the output.
HTML is fastest. Can preview in HTML, but change to pdf or slides later
output: html_document
#output: pdf_document
#output: ioslides_presentation


### Presentations

There are 4 presentation outputs

1. `beamer_presentation`: PDF
1. ` ioslides_presentation`: HTML.useful for sharing in  screen sharing remote meetings
1. `powerpoint_presentation`: Microsoft powerpoint
1. `slidy_presentation`: HTML

Each second level header marks a new slide

Use manual line breaks and insert `***` before each third level header, eg

## Add a table of contents

This is done in YAML, 

output: 
 html_document:
    toc: true
    
need the line breaks and the : !!!!

## R Markdown cheatsheet

select `Help > Cheatsheets > R Markdown Cheat Sheet` from the menu.




    
